hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/MergeIntoHoodieTableCommand.scala

```diff
@@ -25,265 +25,266 @@ import org.apache.hudi.common.model.HoodieAvroRecordMerger
+import org.apache.spark.sql.types.{BooleanType, StructField, StructType}
 
 import java.util.Base64
 
 /**
- * The Command for hoodie MergeIntoTable.
- * The match on condition must contain the row key fields currently, so that we can use Hoodie
- * Index to speed up the performance.
+ * Hudi's implementation of the {@code MERGE INTO} (MIT) Spark SQL statement.
  *
- * The main algorithm:
+ * NOTE: That this implementation is restricted in a some aspects to accommodate for Hudi's crucial
+ *       constraint (of requiring every record to bear unique primary-key): merging condition ([[mergeCondition]])
+ *       is currently can only (and must) reference target table's primary-key columns (this is necessary to
+ *       leverage Hudi's upserting capabilities including Indexes)
  *
- * We pushed down all the matched and not matched (condition, assignment) expression pairs to the
- * ExpressionPayload. And the matched (condition, assignment) expression pairs will execute in the
- * ExpressionPayload#combineAndGetUpdateValue to compute the result record, while the not matched
- * expression pairs will execute in the ExpressionPayload#getInsertValue.
+ * Following algorithm is applied:
  *
- * For Mor table, it is a litter complex than this. The matched record also goes through the getInsertValue
- * and write append to the log. So the update actions & insert actions should process by the same
- * way. We pushed all the update actions & insert actions together to the
- * ExpressionPayload#getInsertValue.
+ * <ol>
+ *   <li>Incoming batch ([[sourceTable]]) is reshaped such that it bears correspondingly:
+ *   a) (required) "primary-key" column as well as b) (optional) "pre-combine" column; this is
+ *   required since MIT statements does not restrict [[sourceTable]]s schema to be aligned w/ the
+ *   [[targetTable]]s one, while Hudi's upserting flow expects such columns to be present</li>
  *
+ *   <li>After reshaping we're writing [[sourceTable]] as a normal batch using Hudi's upserting
+ *   sequence, where special [[ExpressionPayload]] implementation of the [[HoodieRecordPayload]]
+ *   is used allowing us to execute updating, deleting and inserting clauses like following:</li>
+ *
+ *     <ol>
+ *       <li>All the matched {@code WHEN MATCHED AND ... THEN (DELETE|UPDATE ...)} conditional clauses
+ *       will produce [[(condition, expression)]] tuples that will be executed w/in the
+ *       [[ExpressionPayload#combineAndGetUpdateValue]] against existing (from [[targetTable]]) and
+ *       incoming (from [[sourceTable]]) records producing the updated one;</li>
+ *
+ *       <li>Not matched {@code WHEN NOT MATCHED AND ... THEN INSERT ...} conditional clauses
+ *       will produce [[(condition, expression)]] tuples that will be executed w/in [[ExpressionPayload#getInsertValue]]
+ *       against incoming records producing ones to be inserted into target table;</li>
+ *     </ol>
+ * </ol>
+ *
+ * TODO explain workflow for MOR tables
```

Comment:
Changes to this command were required due to 

 - Switching from bespoke resolution to Spark's standard one: we need to abide by Spark's semantic and had to get rid of some customizations implemented previously
 - Cleaned up the code considerably simplifying the implementation by getting rid of custom utilities and replacing them w/ Spark's standard ones (for ex, to resolve, bind expressions, etc)
 - Adding documentation to elaborate on the overall workflow